tools/rewriter/GenCallAsm.cpp

```diff
@@ -387,241 +750,41 @@ std::string emit_asm_wrapper(const CAbiSignature &sig,
-  // so use r10 and r11 as scratch registers
-  add_comment_line(aw,
-                   "Set PKRU to the intermediate value to move return value");
-  add_asm_line(aw, "movq %rax, %r10");
-  add_asm_line(aw, "movq %rdx, %r11");
-  // Change pkru to the intermediate value. This uses rax, r10 and r11 as
-  // scratch registers.
-  emit_mixed_wrpkru(aw, caller_pkey, target_pkey);
-  add_asm_line(aw, "movq %r10, %rax");
-  add_asm_line(aw, "movq %r11, %rdx");
+  emit_set_pkru(aw, target_pkey, kind, arch);
 
-  // Free stack space used for stack args on the target stack
-  if (stack_arg_size > 0) {
-    add_comment_line(aw, "Free stack space used for stack args");
-    add_asm_line(aw, "addq $"s + std::to_string(stack_arg_size) + ", %rsp");
-  }
+  emit_fn_call(target_name, kind, aw, arch);
 
-  // If we inserted 8 bytes to align the stack before calling the wrapped
-  // function, free this space.
-  if (stack_alignment != 0) {
-    assert(stack_alignment == 8);
-    add_asm_line(aw, "addq $8, %rsp");
-  }
+  emit_return_intermediate_pkru(aw, caller_pkey, target_pkey, arch);
 
-  // Copy any stack returns to caller's stack
-  if (stack_return_size > 0) {
-    // If the return value is in memory, we pushed the initial rdi (which
-    // pointed to the caller's return value memory) onto the compartment's
-    // stack. Now we pop it into rax which is where the address of the return
-    // value must be on return.
-    add_asm_line(aw, "popq %rax");
-
-    // After the pop rsp points to the memory for the return value on the
-    // compartment's stack.
-    add_comment_line(aw,
-                     "Copy "s + std::to_string(stack_return_size) +
-                         " bytes for the return value to the caller's stack");
-    for (int i = 0; i < stack_return_size; i += 8) {
-      add_asm_line(aw, "popq "s + std::to_string(i) + "(%rax)");
-    }
+  emit_free_stack_space(aw, stack_arg_size, stack_alignment, arch);
 
-    if (stack_return_padding > 0) {
-      add_asm_line(aw, "addq $8, %rsp");
-    }
-  }
+  emit_copy_stack_returns(aw, stack_return_size, stack_return_padding, caller_pkey, target_pkey, arch);
 
-  // Switch back to the caller's stack
-  emit_switch_stacks(aw, target_pkey, caller_pkey, "r11"s);
+  emit_switch_stacks(aw, target_pkey, caller_pkey, arch);
 
-  // After calling the target function, we only need to zero out registers not
-  // used for return values if the target is a protected compartment
-  if (target_pkey != 0) {
-    add_comment_line(
-        aw, "Push return regs to caller's stack before scrubbing registers");
-    // Push return regs to the caller's stack. These pushes have matching pops
-    // before the return so it has no net effect on the caller's stack pointer.
-    for (auto loc : return_locs) {
-      if (!loc.is_stack()) {
-        emit_reg_push(aw, loc);
-      }
-    }
+  emit_push_return_regs_to_caller_stack(aw, target_pkey, return_locs, arch);
```

Comment:
This and the next two functions (`emit_return_zero_regs` and `emit_pop_return_regs`) should probably be combined into one function. I was thinking something analogous to `emit_zero_regs` so it'll be easier to deduplicate some code in a later PR.

Reply:
Done.